### DISCUSSION TITLE
[Feature]: Move checks for asset upload status into server endpoint

### DISCUSSION THREAD

**Discussion by bo0tzz** [immich-app/immich#1658] [Feature Request]
Currently, clients each implement their own logic to decide whether they should upload an asset to the server. After upload, the server then examines the asset again to determine whether it was indeed appropriate to be uploaded. This places copies of the relevant code in multiple places. Instead, the server should expose an endpoint for clients to ask whether a given asset should be uploaded. With that approach, all the relevant code for supported mimetypes and such can exist only serverside.

This endpoint should take an array of request objects with metadata about the asset. Some optional?, some required!. I propose the following fields:
* `id!` - A unique id by which the client can identify the asset. Used only in the response to this request. (eg `ce784cdd-f459-4d94-b557-9dd73d46cc56`)
* `extension!` - The extension of the asset file (eg `.jpeg`)
* `mimetype!` - The mimetype of the asset file, as determined by the browser or a mimetype library (eg `image/jpeg`)
* `sha1?` - The sha1 hash of the file (`93729ce956531f38867f78858023882147cdfe00`)
* `deviceId?` - The id of the client device (same as current deviceId, maybe for bwcompat?)
* `deviceAssetId?` - The id of the file as determined by the client platform (same as current deviceAssetId, maybe for bwcompat?)

In response, the server should send an array with one object for each of the request objects, stating what action the client should take with that asset. Suggested:
* `id!` - The unique ID as sent by the client in the request, to identify the asset
* `action!` - The action to take on this asset. `Accept`, `Reject` (naming tbd? Can also be `Upload`, `Ignore`, etc)
* `reason?` - If the asset is rejected, the reason why (`Unsupported`, `Duplicate`, etc)

Here's an example request and response:
`POST /api/v1/asset/prepareUpload`
```json
[
  {
    "id": "ce784cdd-f459-4d94-b557-9dd73d46cc56",
    "extension": "png",
    "mimetype": "image/png",
    "sha1": "93729ce956531f38867f78858023882147cdfe00"
  },
  {
    "id": "0fc8024b-43fb-436a-8ced-609268858094",
    "extension": "mp3",
    "mimetype": "audio/mpeg"
  },
  {
    "id": "this can be literally any value",
    "extension": "webp",
    "mimetype": "image/webp",
    "sha1": "5e94be3495d080eed4ca8ebe1fc10b2410625d8d"
  }
]
```
Response:
```json
[
  {
    "id": "ce784cdd-f459-4d94-b557-9dd73d46cc56",
    "action": "Accept"
  },
  {
    "id": "0fc8024b-43fb-436a-8ced-609268858094",
    "action": "Reject",
    "reason": "Unsupported"
  },
  {
    "id": "this can be literally any value",
    "action": "Reject",
    "reason": "Duplicate"
  }
]
```

**Comment by stevenwcarter** [comment]
Instead of calculating sha1 hashes, would it be better to use a hashing algorithm like OSHash that hashes much more quickly? It only hashes the first and last 64K of a file, along with the file's length. https://trac.opensubtitles.org/projects/opensubtitles/wiki/HashSourceCodes

I also wonder if perceptual hashing has been considered for deduplication (though that is only tangentially related to this discussion): https://en.wikipedia.org/wiki/Perceptual_hashing

**Comment by aisbergde** [comment]
perceptual hashing would be fine if embedded metadata was changed, and it will not be detected as duplicate via exact hashing. In this case, a user action would be fine to let him decide, what to do. And some options regarding this.

**Comment by MrModest** [comment]
Can you please link the fix PR or version here as well?

**Reply by bo0tzz** [reply]
I don't have it on hand, this has been implemented for quite a while now we just forgot to close it.

**Reply by MrModest** [reply]
Does it mean, that now if remove Immich application from my phone and install it again, it won't try to upload my every photo from the phone, but just check hashsum/metadata instead?

**Reply by MrModest** [reply]
I see, for example, that the branch [hash-on-upload](https://github.com/immich-app/immich/tree/hash-on-upload) that @alextran1502 mentioned [here](https://github.com/immich-app/immich/pull/9306#issuecomment-2346565973) is still alive with no progress since September 5 last year 🤔

**Reply by bo0tzz** [reply]
No, that's tracked in #2567
